source/common/ssl/ssl_socket.cc

```diff
@@ -227,6 +228,30 @@ std::string SslSocket::sha256PeerCertificateDigest() {
   return Hex::encode(computed_hash);
 }
 
+std::string SslSocket::urlEncodedPemEncodedPeerCertificate() const {
+  bssl::UniquePtr<X509> cert(SSL_get_peer_certificate(ssl_.get()));
+  if (!cert) {
+    return "";
+  }
+
+  bssl::UniquePtr<BIO> buf(BIO_new(BIO_s_mem()));
+  RELEASE_ASSERT(buf != nullptr);
+  RELEASE_ASSERT(PEM_write_bio_X509(buf.get(), cert.get()) == 1);
+  const uint8_t* output;
+  size_t length;
+  RELEASE_ASSERT(BIO_mem_contents(buf.get(), &output, &length) == 1);
+  std::string pem = std::string(reinterpret_cast<const char*>(output), length);
+  // URL encoding shortcut
+  std::vector<std::pair<const absl::string_view, std::string>> replacements;
```

Comment:
I would use initializer list init here and make the variable const. Something like:
```
const std::vector<std::pair<const absl::string_view, std::string>> replacements{{"\n", "%0A"}, ...};
```